ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/transaction/UtxoTransactionBuilder.kt

```diff
@@ -0,0 +1,228 @@
+     *
+     * @param command The command to add to the current [UtxoTransactionBuilder].
+     * @param signatories The associated signatories to add to the current [UtxoTransactionBuilder].
+     * @return Returns a [UtxoTransactionBuilder] including the additional command and associated signatories.
+     */
+    fun addCommandAndSignatories(command: Command, vararg signatories: PublicKey): UtxoTransactionBuilder {
+        return addCommandAndSignatories(command, signatories.toSet())
+    }
+
+    /**
+     * Adds an input state to the current [UtxoTransactionBuilder].
+     *
+     * @param stateAndRef The [StateAndRef] of the input state to add to the current [UtxoTransactionBuilder].
+     * @return Returns a [UtxoTransactionBuilder] including the additional input state.
+     */
+    fun addInputState(stateAndRef: StateAndRef<*>): UtxoTransactionBuilder
+
+    /**
+     * Adds a reference input state to the current [UtxoTransactionBuilder].
+     *
+     * @param stateAndRef The [StateAndRef] of the reference input state to add to the current [UtxoTransactionBuilder].
+     * @return Returns a [UtxoTransactionBuilder] including the additional reference input state.
+     */
+    fun addReferenceInputState(stateAndRef: StateAndRef<*>): UtxoTransactionBuilder
+
+    /**
+     * Adds an output state to the current [UtxoTransactionBuilder].
+     *
+     * @param transactionState The [TransactionState] to add to the current [UtxoTransactionBuilder].
+     * @return Returns a [UtxoTransactionBuilder] including the additional output state.
+     */
+    fun addOutputState(transactionState: TransactionState<*>): UtxoTransactionBuilder
+
+    /**
+     * Adds an output state to the current [UtxoTransactionBuilder].
+     *
+     * @param contractState The [ContractState] to add to the current [UtxoTransactionBuilder].
+     * @return Returns a [UtxoTransactionBuilder] including the additional output state.
+     */
+    fun addOutputState(contractState: ContractState): UtxoTransactionBuilder
+
+    /**
+     * Adds an output state to the current [UtxoTransactionBuilder].
+     *
+     * @param contractState The [ContractState] to add to the current [UtxoTransactionBuilder].
+     * @param notary The notary that will be used to notarise the specified [ContractState].
+     * @return Returns a [UtxoTransactionBuilder] including the additional output state.
+     */
+    fun addOutputState(contractState: ContractState, notary: Party): UtxoTransactionBuilder
+
+    /**
+     * Adds an output state to the current [UtxoTransactionBuilder].
+     *
+     * @param contractState The [ContractState] to add to the current [UtxoTransactionBuilder].
+     * @param contractId The class name of the [Contract] associated with the transaction state.
+     * @return Returns a [UtxoTransactionBuilder] including the additional output state.
+     */
+    fun addOutputState(contractState: ContractState, contractId: String): UtxoTransactionBuilder
```

Comment:
I would be keen to settle on either annotation based linkage to contracts, or a parameter (clazz, not string?) in builder. Not sure what having both implies.

Reply:
@mnesbit I was always skeptical about the `BelongsToContract` annotation, because it means that a state only works with one contract. Before, I'm sure it was possible to use states with different contracts; for example:

```kotlin
data class Person(val owner: PublicKey, val name: String, val birthday: Date) : ContractState {
    val participants: Set<PublicKey> get() = setOf(owner)
}

class UpdatableNameContract : Contract {
    // Impl allows create, update (name only) and delete
}

class UpdatableBirthdayContract : Contract {
    // Impl allows create, update (birthday only) and delete
}
```

From the perspective of transaction construction, once your state is created and bound to a particular contract, you have to stick with it until the state is consumed:

```kotlin
val builder = TransactionBuilder(notary)
    .addOutputState(person, UpdatableNameContract::class)
    .addCommand(UpdatableNameContract.Create, key);
```

In a subsequent transaction, you would not be able to switch contract:
```kotlin
val builder = TransactionBuilder(input.state.notary)
    .addInputState(input)
    .addOutputState(person, UpdatableBirthdayContract.class) // NOT ALLOWED!!!
    .addOutputState(person, UpdatableNameContract.class)
    .addCommand(UpdatableNameContract.UpdateName, key);
```

So, once you've bound your state to a particular contract type, that binding must be applied for all subsequent transactions using that state until consumption, but the important thing is, you can choose different bindings at creation, which lends itself to code reuse.

I'd be inclined to remove the annotation, favouring this...
```kotlin
fun addOutputState(state: ContractState, type: Class<out Contract>): UtxoTransactionBuilder
```

That do you think?

Reply:
Needs a short design meeting I think.

Reply:
@rick-r3 and I drilled into C4 and we have concluded that the versions which pass in strings are possibly unsafe allowing people to hijack states with other contracts. The annotation based linkage between states and the contract class is possible to make safe. I would like to hear if there are any reasons people use the string approach that I haven't heard of, but we should probably kill all the forms taking strings.